File: backend/app/models/asset.py
from enum import Enum
from typing import Optional
from datetime import date, datetime, timezone
from pydantic import BaseModel, Field, ConfigDict


class AssetType(str, Enum):
    SERVER = "Server"
    SWITCH = "Switch"
    ROUTER = "Router"
    FIREWALL = "Firewall"
    LOAD_BALANCER = "Load Balancer"
    STORAGE = "Storage"
    VM = "Virtual Machine"
    OTHER = "Other"


class AssetStatus(str, Enum):
    ACTIVE = "Active"
    INACTIVE = "Inactive"
    MAINTENANCE = "Maintenance"
    DECOMMISSIONED = "Decommissioned"


class Asset(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: Optional[str] = Field(default=None, alias="_id")
    name: str
    asset_type: AssetType
    status: AssetStatus = AssetStatus.ACTIVE
    ip_record_id: Optional[str] = None
    hostname: Optional[str] = None
    serial_number: Optional[str] = None
    vendor: Optional[str] = None
    model: Optional[str] = None
    os_version: Optional[str] = None
    data_center: Optional[str] = None
    rack_location: Optional[str] = None
    warranty_expiry: Optional[date] = None
    notes: Optional[str] = None
    tags: list[str] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    created_by: str = "system"
    updated_by: str = "system"

File: frontend/src/types/asset.ts
export type AssetType =
  | 'Server'
  | 'Switch'
  | 'Router'
  | 'Firewall'
  | 'Load Balancer'
  | 'Storage'
  | 'Virtual Machine'
  | 'Other';

export type AssetStatus = 'Active' | 'Inactive' | 'Maintenance' | 'Decommissioned';

export interface Asset {
  id: string;
  name: string;
  asset_type: AssetType;
  status: AssetStatus;
  ip_record_id: string | null;
  ip_address: string | null;
  hostname: string | null;
  serial_number: string | null;
  vendor: string | null;
  model: string | null;
  os_version: string | null;
  data_center: string | null;
  rack_location: string | null;
  warranty_expiry: string | null; // ISO date string
  notes: string | null;
  tags: string[];
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export interface AssetCreate {
  name: string;
  asset_type: AssetType;
  status?: AssetStatus;
  ip_record_id?: string | null;
  hostname?: string | null;
  serial_number?: string | null;
  vendor?: string | null;
  model?: string | null;
  os_version?: string | null;
  data_center?: string | null;
  rack_location?: string | null;
  warranty_expiry?: string | null;
  notes?: string | null;
  tags?: string[];
}

export interface AssetUpdate {
  name?: string;
  asset_type?: AssetType;
  status?: AssetStatus;
  ip_record_id?: string | null;
  hostname?: string | null;
  serial_number?: string | null;
  vendor?: string | null;
  model?: string | null;
  os_version?: string | null;
  data_center?: string | null;
  rack_location?: string | null;
  warranty_expiry?: string | null;
  notes?: string | null;
  tags?: string[];
}

export interface AssetFilters {
  page?: number;
  page_size?: number;
  asset_type?: AssetType;
  status?: AssetStatus;
  data_center?: string;
  search?: string;
}

export const ASSET_TYPES: AssetType[] = [
  'Server',
  'Switch',
  'Router',
  'Firewall',
  'Load Balancer',
  'Storage',
  'Virtual Machine',
  'Other',
];

export const ASSET_STATUSES: AssetStatus[] = [
  'Active',
  'Inactive',
  'Maintenance',
  'Decommissioned',
];
